transformer_engine/common/transpose/cast_transpose.cu

```diff
@@ -6,432 +6,360 @@
-                           warp_id_in_tile * n_iterations) %
-                         THREADS_PER_WARP;
-  CType max = 0;
-  const CType scale = scale_ptr != nullptr ? *scale_ptr : 1;
-#pragma unroll
-  for (unsigned int i = 0; i < nvec_out; ++i) {
-    in[0][i].load_from(my_input_tile, current_stride + my_place + stride * i);
-  }
-#pragma unroll
-  for (unsigned int i = 0; i < n_iterations; ++i) {
-    const size_t current_place = current_stride + my_place;
-    const unsigned int my_place_in = (my_place + THREADS_PER_WARP - 1) % THREADS_PER_WARP;
-    const unsigned int current_in = (i + 1) % 2;
-    if (i < n_iterations - 1) {
-#pragma unroll
-      for (unsigned int j = 0; j < nvec_out; ++j) {
-        in[current_in][j].load_from(my_input_tile,
-                                    current_stride + my_place_in + stride * (nvec_out + j));
-      }
+namespace {
+
+// String with RTC kernel implementation
+#include "string_code_transpose_rtc_cast_transpose_cu.h"
+
+// Hard-coded kernel parameters
+using CType = float;
+constexpr size_t warps_per_tile = 4;
+constexpr size_t block_size = THREADS_PER_WARP * warps_per_tile;
+
+/* Performance heuristics for optimized kernel parameters */
+struct KernelConfig {
+  /** Vector load size */
+  size_t load_size = 0;
+  /** Vector store size to transposed output */
+  size_t store_size = 0;
+
+  /* Whether config is valid */
+  bool valid = false;
+  /* Number of CUDA blocks */
+  size_t num_blocks = 0;
+
+  /* Number of active SMs */
+  size_t active_sm_count = 0;
+  /* Elements per L1 cache load */
+  size_t elements_per_load = 0;
+  /* Elements per L1 cache store to cast output*/
+  size_t elements_per_store_c = 0;
+  /* Elements per L1 cache store to transposed output */
+  size_t elements_per_store_t = 0;
+
+  KernelConfig(size_t row_length,
+               size_t num_rows,
+               size_t itype_size,
+               size_t otype_size,
+               size_t load_size_,
+               size_t store_size_)
+    : load_size{load_size_}
+    , store_size{store_size_} {
+    // Check that tiles are correctly aligned
+    constexpr size_t cache_line_size = 128;
+    if (load_size % itype_size != 0
+        || store_size % otype_size != 0
+        || cache_line_size % itype_size != 0
+        || cache_line_size % otype_size != 0) {
+      return;
     }
-    OVec out_trans[nvec_in];  // NOLINT(*)
-    cast_and_transpose_regs<true>(in[current_in ^ 1], out_trans, my_output_c_tile,
-                                  current_place, stride, max, scale, true);
-#pragma unroll
-    for (unsigned int j = 0; j < nvec_in; ++j) {
-      out_space[i][j].data.vec = out_trans[j].data.vec;
+    const size_t row_tile_elements = load_size * THREADS_PER_WARP / itype_size;
+    const size_t col_tile_elements = store_size * THREADS_PER_WARP / otype_size;
+    valid = (row_length % row_tile_elements == 0
+             && num_rows % col_tile_elements == 0);
+    if (!valid) {
```

Comment:
Maybe to add a notice that the object has not been fully constructed and might be in inconsistent state?